gen3-client/g3cmd/upload-multipart.go

```diff
@@ -0,0 +1,172 @@
+package g3cmd
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"os"
+	"sort"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/uc-cdis/gen3-client/gen3-client/commonUtils"
+	"github.com/uc-cdis/gen3-client/gen3-client/logs"
+	pb "gopkg.in/cheggaaa/pb.v1"
+)
+
+var multipartUploadLock sync.Mutex
+
+func retry(attempts int, filePath string, guid string, f func() error) (err error) {
+	for i := 0; ; i++ {
+		err = f()
+		if err == nil {
+			return
+		}
+
+		if i >= (attempts - 1) {
+			break
+		}
+
+		time.Sleep(GetWaitTime(i))
+
+		log.Println("Retrying after error: ", err)
+	}
+	return fmt.Errorf("After %d attempts, last error: %s", attempts, err)
+}
+
+func multipartUpload(uploadPath string, filePath string, includeSubDirName bool, retryCount int) error {
+	file, err := os.Open(filePath)
+	defer file.Close()
+	if err != nil {
+		logs.AddToFailedLogMap(filePath, "", retryCount, true, true)
+		err = fmt.Errorf("FAILED multipart upload for %s due to file open error: %s", filePath, err.Error())
+		return err
+	}
+
+	fi, err := file.Stat()
+	if err != nil {
+		logs.AddToFailedLogMap(filePath, "", retryCount, true, true)
+		err = fmt.Errorf("FAILED multipart upload for %s: file stat error, file may be missing or unreadable because of permissions", fi.Name())
+		return err
+	}
+
+	if fi.Size() > MultipartFileSizeLimit {
+		logs.AddToFailedLogMap(filePath, "", retryCount, true, true)
+		err = fmt.Errorf("FAILED multipart upload for %s: the file size has exceeded the limit allowed and cannot be uploaded. The maximum allowed file size is 5TB", fi.Name())
```

Comment:
Should this print use MultipartFileSizeLimit in the string instead of hardcoding 5TB?

Reply:
good point! fixed in https://github.com/uc-cdis/cdis-data-client/pull/50/commits/5fbbda24b80735667210e37c9869bd64f8bc46c9